### ISSUE TITLE
boost::core::bit_width not constexpr in VS

### ISSUE THREAD

**Issue by joaquintides** [boostorg/core#109]
The [following](https://godbolt.org/z/oeocqq3bo)
```cpp
#include <boost/core/bit.hpp>

int x[boost::core::bit_width(16u)];
```
yields
```
<source>(3): error C3615: constexpr function 'boost::core::bit_width' cannot result in a constant expression
C:/data/libraries/installed/x64-windows/include\boost/core/bit.hpp(445): note: failure was caused by call of undefined function or one not declared 'constexpr'
C:/data/libraries/installed/x64-windows/include\boost/core/bit.hpp(445): note: see usage of 'boost::core::countl_zero'
ASM generation compiler returned: 2
```

**Comment by pdimov** [comment]
Yes, because the MSVC bitscan intrinsics aren't constexpr.

**Comment by joaquintides** [comment]
Can't `std::is_constant_evaluated()` be leveraged to remedy this?

**Comment by pdimov** [comment]
`std::is_constant_evaluated` is C++20, so the gains will be marginal.

**Comment by joaquintides** [comment]
This is better than nothing... At least the `BOOST_CONSTEXPR` qualifier should be conditionally removed in VS.

**Comment by pdimov** [comment]
There's no need to remove the qualifier.

It might be possible to use the is_constant_evaluated intrinsic: https://github.com/microsoft/STL/blob/5e3574b97985db31112a1c60e4eb52e821d25a83/stl/inc/xtr1common#L170

This should be available in all language modes, but I'll need to check which MSVC versions have it.

**Comment by joaquintides** [comment]
Looks like the intrinsic is available from [Visual Studio 2019, versión 16.5](https://godbolt.org/z/ds3s8qGz1) onwards.
